### PR TITLE
PropEr testing

### PR DESCRIPTION
**Pull request by jamesaimonetti**
Creates a framework for building PropEr based testing of Crossbar APIs, with basic testing of `/phone_numbers` as a start.

### CODE REVIEW

File: applications/crossbar/src/modules/cb_search.erl [hunk 1]
```diff
@@ -411,7 +410,7 @@ get_end_key(Context, Type, Value) ->
 %%--------------------------------------------------------------------
 -spec next_binary_key(binary()) -> ne_binary().
 next_binary_key(<<>>) ->
-    <<"\ufff0">>;
+    <<0>>;
```
Comment (fenollp) [inline applications/crossbar/src/modules/cb_search.erl]: Are you sure about this one? It's supposed to be the biggest possible string. Isn't `<<0>>` considered the smallest?
Actually is this valid?

File: core/kazoo_number_manager/src/kazoo_number_manager_maintenance.erl [hunk 2]
```diff
@@ -212,10 +212,9 @@ update_number_services_view(?MATCH_ACCOUNT_ENCODED(_)=AccountDb) ->
                                 ,View
                                 ),
     case kz_json:are_equal(View, NewView) of
-        true -> ?LOG("View is up to date.", []);
+        true -> 'ok';
         false ->
-            true = kz_datamgr:db_view_update(AccountDb, [{ViewName, NewView}]),
-            ?LOG("View updated!", [])
```
Comment (fenollp) [inline core/kazoo_number_manager/src/kazoo_number_manager_maintenance.erl]: Please keep the above log line. Helps a lot.

File: core/kazoo_proper/src/pqc_cb_phone_numbers.erl [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+                        'true' -> 'reserved' =:= pqc_cb_response:number_state(APIResult);
+                        'false' -> 500 =:= pqc_cb_response:error_code(APIResult)
+                    end
+            end
+    end;
+postcondition(Model
+             ,{'call', _, 'remove_number', [_API, AccountId, Number]}
+             ,APIResult
+             ) ->
+    case pqc_kazoo_model:does_account_exist(Model, AccountId) of
```
Comment (fenollp) [inline core/kazoo_proper/src/pqc_cb_phone_numbers.erl]: If Model was a map() or if we could use macros to match this property from the model, pre/post-condition code would get much less lengthier. What do you think? I believe that with time we would be checking more and more properties and these clauses will get monstruous.

File: core/kazoo_stdlib/test/kz_json_test.erl [hunk 4]
```diff
@@ -19,21 +19,28 @@
...
+     ,[{atom_to_list(F)
+       ,fun () ->
+                ?assert(proper:quickcheck(?MODULE:F(), [{'to_file', 'user'}
+                                                       ,{'numtests', 500}
+                                                       ]))
+        end
+       }
+       || {F, 0} <- ?MODULE:module_info('exports'),
+          F > 'prop_',
+          F < 'prop`'
```
Comment (fenollp) [inline core/kazoo_stdlib/test/kz_json_test.erl]: Pretty sure this is not doing what you think it's doing: needs conversion to a string type.

File: core/kazoo_proper/src/pqc_cb_phone_numbers.erl (hunk 3 above)
Comment (jamesaimonetti) [inline core/kazoo_proper/src/pqc_cb_phone_numbers.erl]: Agreed. I tried to keep it simple to make sure failing runs could be debugged easily while I worked out the kinks (hence one account, one number for the time being). Not married to how predicates are defined for post-conditions or next_state transitions; will let them evolve as this moves forward.

File: core/kazoo_stdlib/test/kz_json_test.erl (hunk 4 above)
Comment (jamesaimonetti) [inline core/kazoo_stdlib/test/kz_json_test.erl]: Missed the '`' in the second one to get the sorting. Worked in the shell, copy/pasted wrong apparently.

**Comment by jamesaimonetti** [comment]
@fenollp At the moment, this runs locally. Enter the master account API key, compile, and away you go. The README in `doc/` should have the sequential and parallel test initiation commands to run in the shell. I wanted to get some eyes on it and extend it a bit more before setting up any CI testing.

**Comment by jamesaimonetti** [comment]
@fenollp also, with respect to the SDK, that would be great to have.

There's going to be trickiness once we start testing POST and others that require the data from a previous GET/PUT. Numbers was easy with mostly PUTs and DELETEs that didn't require tracking an ID. I was going to do `/accounts` next and see what types of race conditions there lurking in name/realm uniqueness!

**Comment by fenollp** [comment]
Cool!
Yes trickiness will ensue but it's definitely doable. It's a state machine after all :)

**Comment by lazedo** [comment]
@fenollp  what changes are you requiring here ?

**Comment by fenollp** [comment]
Just some Dialyzer issues and we're good to go!

**Comment by fenollp** [comment]
@jamesaimonetti props PropEr tests failing

**Comment by jamesaimonetti** [comment]
@fenollp i constrained the tests because we can't distinguish list-as-key vs list-as-path and trying to generate nested proplists got caught in a recursive loop trap or something.

**Pull request merged**